1.16 made JetBrain Gradle plugin mandatory

```
IntelliJ IDEA 2022.3.3 (Community Edition)
Build #IC-223.8836.41, built on March 10, 2023
Runtime version: 17.0.6+1-b653.34 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Windows 10 10.0
GC: G1 Young Generation, G1 Old Generation
Memory: 2018M
Cores: 12
Non-Bundled Plugins:
    com.redhat.devtools.intellij.telemetry (0.0.3.33)
    com.redhat.devtools.intellij.quarkus (1.16.0.329)

Kotlin: 223-1.8.0-release-345-IJ8836.41
```

IntelliJ repeatedly shows the following exception while working in both Quarkus and other types of projects when the Quarkus plug-in is enabled:

```
Internal error: java.lang.NoClassDefFoundError: org/jetbrains/plugins/gradle/util/GradleConstants

java.util.concurrent.CompletionException: java.lang.NoClassDefFoundError: org/jetbrains/plugins/gradle/util/GradleConstants
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient.runTask(IndexAwareLanguageClient.java:101)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient$1.run(IndexAwareLanguageClient.java:76)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:429)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressInCurrentThread$11(CoreProgressManager.java:557)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcessWithProgressInCurrentThread(CoreProgressManager.java:557)
	at com.intellij.openapi.progress.impl.CoreProgressManager.run(CoreProgressManager.java:363)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient.lambda$runAsBackground$0(IndexAwareLanguageClient.java:73)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient.lambda$runAsBackground$1(IndexAwareLanguageClient.java:87)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.NoClassDefFoundError: org/jetbrains/plugins/gradle/util/GradleConstants
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.GradleProjectLabelProvider.isGradleProject(GradleProjectLabelProvider.java:45)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.GradleProjectLabelProvider.getProjectLabels(GradleProjectLabelProvider.java:38)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager.getProjectLabels(ProjectLabelManager.java:122)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager.getProjectLabelInfo(ProjectLabelManager.java:77)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager.lambda$getProjectLabelInfo$1(ProjectLabelManager.java:102)
	at com.intellij.openapi.project.DumbService.lambda$runReadActionInSmartMode$0(DumbService.java:112)
	at com.intellij.openapi.project.DumbService.lambda$runReadActionInSmartMode$1(DumbService.java:156)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:898)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:68)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.java:149)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.java:112)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager.getProjectLabelInfo(ProjectLabelManager.java:102)
	at com.redhat.devtools.intellij.quarkus.lsp.QuarkusLanguageClient.lambda$getJavaProjectlabels$6(QuarkusLanguageClient.java:132)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient.runTask(IndexAwareLanguageClient.java:97)
	... 22 more
Caused by: java.lang.ClassNotFoundException: org.jetbrains.plugins.gradle.util.GradleConstants PluginClassLoader(plugin=PluginDescriptor(name=Quarkus Tools, id=com.redhat.devtools.intellij.quarkus, descriptorPath=plugin.xml, path=~\AppData\Roaming\JetBrains\IdeaIC2022.3\plugins\Quarkus Tools, version=1.16.0.329, package=null, isBundled=false), packagePrefix=null, instanceId=33, state=active)
	... 36 more
```

The package `org.jetbrains.plugins.gradle` is in the JetBrains Gradle plug-in. But even if I enabled this plug-in the exception still occurs frequently.

The Quarkus plug-in declares an optional dependency on the JetBrains Gradle plug-in. It is using the above package name in the declaration in plugin.xml. But according to the JetBrains marketplace the plug-in id for the JetBrains Gradle plug-in is `com.intellij.gradle`.

With the change of plug-in id the exceptions ceases when the JetBrains Gradle plug-in is enabled. But it still occurs repeatedly when JetBrains Gradle plug-in is disabled.

To make 1.16 again treat JetBrains Gradle plug-in as optional, the `GradleProjectLabelProvider` needs to check if the JetBrains Gradle plug-in is enabled. I'm a bit surprised that the check can be in the `GradleProjectLabelProvider` class having the import from the JetBrains Gradle plug-in class. It looks like Java can defer triggering the `java.lang.NoClassDefFoundError` exception until the code actually accesses the missing class.

For above changes please refer to the pull request.

One more class `com.redhat.devtools.intellij.quarkus.gradle.AbstractGradleToolDelegate` also refers to the same class in the JetBrains Gradle plug-in as in the exception above. However, this does not look like it would trigger an exception. This simply because `AbstractGradleToolDelegate` class will only be used if the failing code actually succeeded in identifying the project as Gradle. So I assume it is enough to protect the failing code and this should then not need the same plug-in check.

--- a/src/main/resources/META-INF/plugin-gradle.xml
+++ b/src/main/resources/META-INF/plugin-gradle.xml
@@ -3,6 +3,7 @@
   <extensions defaultExtensionNs="com.redhat.devtools.intellij.quarkus">
     <toolDelegate implementation="com.redhat.devtools.intellij.quarkus.gradle.GradleGroovyToolDelegate"/>
     <toolDelegate implementation="com.redhat.devtools.intellij.quarkus.gradle.GradleKotlinToolDelegate"/>
+    <projectLabelProvider implementation="com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.GradleProjectLabelProvider"/>
   </extensions>
   <extensions defaultExtensionNs="com.intellij">
     <externalProjectDataService implementation="com.redhat.devtools.intellij.quarkus.gradle.QuarkusProjectDataService"/>

--- a/src/main/resources/META-INF/plugin-maven.xml
+++ b/src/main/resources/META-INF/plugin-maven.xml
@@ -2,6 +2,7 @@
 
   <extensions defaultExtensionNs="com.redhat.devtools.intellij.quarkus">
     <toolDelegate implementation="com.redhat.devtools.intellij.quarkus.maven.MavenToolDelegate"/>
+    <projectLabelProvider implementation="com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.MavenProjectLabelProvider"/>
   </extensions>
   <extensions defaultExtensionNs="org.jetbrains.idea.maven">
     <!-- Add your extensions here -->

--- a/src/main/resources/META-INF/plugin.xml
+++ b/src/main/resources/META-INF/plugin.xml
@@ -274,8 +274,6 @@
     <javaDiagnosticsParticipant implementation="com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.metrics.java.MicroProfileMetricsDiagnosticsParticipant"/>
 
     <projectLabelProvider implementation="com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.MicroProfileProjectLabelProvider"/>
-    <projectLabelProvider implementation="com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.MavenProjectLabelProvider"/>
-    <projectLabelProvider implementation="com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.GradleProjectLabelProvider"/>
 
     <javaDefinitionParticipant implementation="com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.config.java.MicroProfileConfigDefinitionParticipant"/>
     <javaDefinitionParticipant implementation="com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.faulttolerance.java.MicroProfileFaultToleranceDefinitionParticipant"/>

